dbms/src/Interpreters/Context.cpp

```diff
@@ -156,7 +156,8 @@ struct ContextShared
     std::optional<SystemLogs> system_logs;                              /// Used to log queries and operations on parts
 
     std::unique_ptr<TraceCollector> trace_collector;        /// Thread collecting traces from threads executing queries
-
+    std::unordered_set<std::string> allowed_primary_url_hosts;      /// Allowed primary (<host>) URL from config.xml
```

Comment:
You can encapsulate it into separate class that will be responsible for:
- initializing from configuration;
- storing the set of allowed hosts;
- checking if the host is allowed;
- throwing exceptions.

It's much better than if it's spread across the codebase.